rs/nns/governance/src/voting.rs

```diff
@@ -20,63 +50,176 @@ impl Governance {
+                    is_voting_finished = machine.is_voting_finished();
+                });
+            });
+            // We send a no-op message to self to break up the call context into more messages
+            noop_self_call_if_over_instructions(
+                SOFT_VOTING_INSTRUCTIONS_LIMIT,
+                Some(HARD_VOTING_INSTRUCTIONS_LIMIT),
+            )
+            .await;
+        }
+    }
+
+    /// Record a neuron vote into the voting state machine, then do nothing else.
+    fn record_neuron_vote(
+        &mut self,
+        proposal_id: ProposalId,
+        voting_neuron_id: NeuronId,
+        vote: Vote,
+        topic: Topic,
+    ) {
+        with_voting_state_machines_mut(|voting_state_machines| {
+            let ballots = &mut self
+                .heap_data
+                .proposals
+                .get_mut(&proposal_id.id)
+                .unwrap()
+                .ballots;
+            voting_state_machines.with_machine(proposal_id, topic, |machine| {
+                machine.cast_vote(ballots, voting_neuron_id, vote)
+            });
+        });
+    }
+
+    /// Process a single voting state machine until it is over the soft limit, then return
+    fn process_machine_until_soft_limit(
+        &mut self,
+        machine: &mut ProposalVotingStateMachine,
+        is_over_soft_limit: fn() -> bool,
+    ) {
+        let proposal_id = machine.proposal_id;
+        while !machine.is_completely_finished() {
+            machine.continue_processing(
+                &mut self.neuron_store,
+                &mut self
+                    .heap_data
+                    .proposals
+                    .get_mut(&proposal_id.id)
+                    .unwrap()
+                    .ballots,
+                is_over_soft_limit,
+            );
+
+            if is_over_soft_limit() {
+                break;
             }
+        }
+    }
+
+    /// Process all voting state machines.  This function is called in the timer job.
+    /// It processes voting state machines until the soft limit is reached.
+    pub fn process_voting_state_machines(&mut self) {
+        let mut proposals_processed = btreeset! {};
+        with_voting_state_machines_mut(|voting_state_machines| loop {
+            if voting_state_machines
+                .with_next_machine(|(proposal_id, machine)| {
+                    // We need to keep track of which proposals we processed
+                    proposals_processed.insert(proposal_id.id);
+                    self.process_machine_until_soft_limit(machine, over_soft_message_limit);
+                })
+                .is_none()
+            {
+                break;
+            };
 
-            voting_state_machines.remove_if_done(&proposal_id);
+            if over_soft_message_limit() {
+                break;
+            }
         });
+        // TODO DO NOT MERGE Is this needed here?  It's also processed
```

Comment:
Would like reviewer attention to this section, I don' think this would hurt, but it may also be redundant since we're doing this in heartbeats already.

Reply:
Second thought - I now believe this should be removed, since it's only for registering recent ballots?

Reply:
Yeah I think it's not needed.